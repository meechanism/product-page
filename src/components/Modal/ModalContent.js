import React from 'react';
import ReactDOM from 'react-dom';
import FocusTrap from 'focus-trap-react';
import styled from 'styled-components';

import Colors from '../../styles/colors';
import MediaQuery from '../../styles/media-queries';

const ModalWrapper = styled.aside.attrs({
  tag: 'aside',
  role: 'dialog',
  tabIndex: -1,
  'aria-modal': true
})`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10;
  transform: translateZ(0);
  background-color: rgba(0, 0, 0, 0.5);
`;

const CloseButton = styled.button.attrs((props) => ({
  'aria-label': 'Close Modal',
  'aria-labelledby': 'close-modal'
}))`
  position: absolute;
  top: 0;
  right: 0;
  padding: 0.5em;
  line-height: 1;
  background: ${Colors.trueWhite};
  border: 0;
  box-shadow: 0;
  cursor: pointer;

  &:hover {
    svg {
      stroke: ${Colors.primary};
    }
  }
`;

const ModalArea = styled.div`
  background-color: ${Colors.trueWhite};
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.1);
  border-radius: 2px;
  position: fixed;
  width: 100%;
  overflow-y: auto;
  overflow-x: hidden;
  -webkit-overflow-scrolling: touch;

  left: 50%;
  top: 50%;
  height: auto;
  transform: translate(-50%, -50%);
  max-width: 30em;
  max-height: calc(100% - 1em);

  ${MediaQuery.small`
    width: 100%;
  `};
`;

const CloseIcon = styled.svg`
  width: 25px;
  height: 25px;
  fill: transparent;
  stroke: black;
  stroke-linecap: round;
  stroke-width: 2;
`;

const CloseVisual = styled.span`
  border: 0 !important;
  clip: rect(0 0 0 0) !important;
  height: 1px !important;
  margin: -1px !important;
  overflow: hidden !important;
  padding: 0 !important;
  position: absolute !important;
  width: 1px !important;
  white-space: nowrap !important;
`;

const ModalContent = ({
  buttonRef,
  onClickOutside,
  onKeyDown,
  modalRef,
  closeModal,
  content
}) => {
  return ReactDOM.createPortal(
    <FocusTrap>
      <ModalWrapper onClick={onClickOutside} onKeyDown={onKeyDown}>
        <ModalArea ref={modalRef}>
          <CloseButton ref={buttonRef} onClick={closeModal}>
            <CloseVisual id="close-modal">Close</CloseVisual>
            <CloseIcon viewBox="0 0 40 40">
              <path d="M 10,10 L 30,30 M 30,10 L 10,30" />
            </CloseIcon>
          </CloseButton>
          <div>{content}</div>
        </ModalArea>
      </ModalWrapper>
    </FocusTrap>,
    document.body
  );
};

export default ModalContent;
