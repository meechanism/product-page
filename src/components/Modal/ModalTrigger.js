import React from 'react';
import Button from '../Button';

export const onClick = (showModal) => (e) => {
  e.preventDefault();
  showModal();
};

const ModalTrigger = ({ buttonRef, showModal, triggerText }) => (
  <Button secondary ref={buttonRef} onClick={onClick(showModal)}>
    {triggerText}
  </Button>
);

export default ModalTrigger;
