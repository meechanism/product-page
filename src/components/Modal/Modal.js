import React, { useState } from 'react';
import ModalContent from './ModalContent';
import ModalTrigger from './ModalTrigger';

export const onKeyDown = (closeFn) => (event) => {
  if (event.keyCode === 27) {
    closeFn();
  }
};

export const toggleScrollLock = () => {
  document.querySelector('html').classList.toggle('scroll-lock');
};

export const showModal = (showFn, scrollLock, closeButton) => () => {
  showFn(true);
  closeButton && closeButton.focus();
  scrollLock();
};

export const closeModal = (showFn, scrollLock, triggerButton) => () => {
  showFn(false);
  triggerButton && triggerButton.focus();
  scrollLock();
};

export const onClickOutside = (closeFn, modal) => (event) => {
  if (modal && modal.contains(event.target)) return;
  closeFn();
};

export const Modal = ({ modalProps, modalContent }) => {
  const [isShown, setIsShown] = useState(false);
  const [modal, setModal] = useState(null);
  const [triggerButton, setTriggerButton] = useState(null);
  const [closeButton, setCloseButton] = useState(null);

  const closeFn = closeModal(setIsShown, toggleScrollLock, triggerButton);
  return (
    <>
      <ModalTrigger
        showModal={showModal(setIsShown, toggleScrollLock, closeButton)}
        buttonRef={(n) => setTriggerButton(n)}
        triggerText={modalProps.triggerText}
      />
      {isShown ? (
        <ModalContent
          modalRef={(n) => setModal(n)}
          buttonRef={(n) => setCloseButton(n)}
          closeModal={closeFn}
          content={modalContent}
          onKeyDown={onKeyDown(closeFn)}
          onClickOutside={onClickOutside(closeFn, modal)}
        />
      ) : null}
    </>
  );
};

export default Modal;
