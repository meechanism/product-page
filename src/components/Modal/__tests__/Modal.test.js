import React from 'react';
import Modal, {
  onKeyDown,
  showModal,
  closeModal,
  onClickOutside,
  toggleScrollLock
} from '../Modal';

import ModalTrigger from '../ModalTrigger';

describe('Modal component', () => {
  let mockFn;
  let mockFn2;
  let mockBtn;

  const TestContent = () => <p>Hello there</p>;
  const props = {
    modalProps: {
      triggerText: 'Open'
    }
  };

  beforeEach(() => {
    mockFn = jest.fn();
    mockFn2 = jest.fn();
    mockBtn = {
      focus: jest.fn()
    };
  });

  it('should render', () => {
    const output = shallow(
      <Modal {...props}>
        <TestContent />
      </Modal>
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('should render trigger', () => {
    const output = shallow(
      <Modal {...props}>
        <TestContent />
      </Modal>
    );
    expect(output.find(ModalTrigger).exists()).toBeTruthy();
  });

  describe('onKeyDown', () => {
    it('should trigger passed in fn when esc button is pressed', () => {
      expect(onKeyDown(mockFn)({ keyCode: 27 }));
      expect(mockFn).toHaveBeenCalled();
    });

    it('should NOT trigger passed in fn when other button is pressed', () => {
      expect(onKeyDown(mockFn)({ keyCode: 11 }));
      expect(mockFn).not.toHaveBeenCalled();
    });
  });

  it('should closeModal', () => {
    closeModal(mockFn, mockFn2, mockBtn)();
    expect(mockFn).toHaveBeenCalledWith(false);
    expect(mockFn2).toHaveBeenCalled();
    expect(mockBtn.focus).toHaveBeenCalled();
  });

  it('should showModal', () => {
    showModal(mockFn, mockFn2, mockBtn)();
    expect(mockFn).toHaveBeenCalledWith(true);
    expect(mockFn2).toHaveBeenCalled();
    expect(mockBtn.focus).toHaveBeenCalled();
  });

  describe('onClickOutside', () => {
    it('should NOT close when area outside modal is clicked', () => {
      const modal = {
        contains: (val) => val
      };
      const e = {
        target: true
      };
      onClickOutside(mockFn, modal)(e);
      expect(mockFn).not.toHaveBeenCalled();
    });

    it('should  close when area outside modal is clicked', () => {
      const modal = {
        contains: (val) => val
      };
      const e = {
        target: false
      };
      onClickOutside(mockFn, modal)(e);
      expect(mockFn).toHaveBeenCalled();
    });
  });

  it('should toggle scroll lock', () => {
    const htmlNode = document.querySelector('html');
    expect(htmlNode.classList.length).toBe(0);
    toggleScrollLock();
    expect(htmlNode.classList.length).toBe(1);
    expect(htmlNode.classList[0]).toBe('scroll-lock');
    toggleScrollLock();
    expect(htmlNode.classList.length).toBe(0);
  });
});
