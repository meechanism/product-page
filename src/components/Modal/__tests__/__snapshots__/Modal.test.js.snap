// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render 1`] = `
<Fragment>
  <ModalTrigger
    buttonRef={[Function]}
    showModal={[Function]}
    triggerText="Open"
  />
</Fragment>
`;
